dependent_types.md

```diff
@@ -0,0 +1,279 @@
+data Either a b = Left a | Right b
+```
+If we rewrote these types with hypotetical syntax we would come to this:
+```
+data Maybe a = Just a
+     Maybe a = Nothing
+
+data Either a b = Left a
+     Either a b = Right b
+```
+and this can inform us that we are really limited with what we do on left side of the `=` sign. 
+
+Problem with this is that we don't get to do pattern matching on the type constructors which in turn means that we can only have free type variables on the left side.
+
+This is where *GADTs* language extension comes to play. It allows us to define type contraints on the type variable which limits the possible type variables that can be used to construct a type and later in the code have *local assumptions* when pattern matching on the data constructor which was used to constuct a type. 
+
+Lets take a look at some example code:
+
+```
+data IntOrString a where
+    IntConstructor    :: a ~ Int    => IntOrString a -- a has a type constraint to Int
+    StringConstructor :: a ~ String => IntOrString a -- a has a type constraint to String
+```
+This is desugared version of the more convenient syntax that we usually use
+
+```
+data IntOrString a where
+    IntConstructor    :: Int    -> IntOrString a
+    StringConstructor :: String -> IntOrString a
+```
+
+The classic `Maybe` datatype would look like this if written with GADTs
+
+```
+data Maybe a where
+    Just    :: a -> Maybe a
+    Nothing :: Maybe a
+```
+And here is example of the pattern match:
+
+```
+wasItStringOrInt :: IntOrString a -> String
+wasItStringOrInt x =
+    case x of
+        -- Local assumptions - Local because it is limited to only one case branch 
+        -- and assumption because we assume what type is our type parameter
+        IntConstructor a    -> "Constructed with Int"      -- a ~ Int
+        StringConstructor b -> "Constructed with String    -- a ~ String
+        
+-- ghci
+λ> let x = IntConstructor 42
+λ> wasItStringOrInt x
+"Constructed with Int"
+```
+
+## Generic
+
+Lets now build our intuition about types in general and also dependent types.
+
+If we look at a normal Haskell Algebraic Data Type `T`
+
+```
+data T = A Int Int
+  | B String String String
+  | C () Void
+  | D
+```
+
+We can say that `T` is a sum of products `A B C D`. In order to understand sum we can define this and every other ADT only using
+`Either`, `() (unit)` , `(,) tuple`  and function type. So our ADT could be encoded like this using infix notation
+
+```
+type T' = (Int, Int) `Either` (String, String, String) `Either` ((), Void) `Either` ()
+```
+
+Tuples provide us with products, Either provides us with sums, unit is empty constructor and function type is just a built in function. 
+
+If we understand these simple types than we understand Algebraic Data Types.
+
+If we understand Σ (sigma) and Π (pi) we understand dependent types. 
+
+You can see the theory behing all of this [here](https://en.wikipedia.org/wiki/Intuitionistic_type_theory) .
```

Comment:
Do you expect the reader to follow the link and get confused with all the technical stuff there? The link is useful, but probably in a sidenote.